#12081: "Add deleteFileThreshold parameter to SizeBasedDataRewriter, update logic, and include tests"

jangalasriramd7 opened the pull request:
This PR refines the SizeBasedDataRewriter class in the Apache Iceberg project, addressing issues related to the constructor initialization and logic for delete file threshold handling. It makes the class more robust by improving its flexibility and functionality in the context of delete-based data file rewriting.

Constructor Updates:
Refactored constructors to properly initialize the deleteRatioThreshold and deleteFileThreshold parameters.
Added an additional constructor that allows specifying the deleteRatioThreshold value explicitly while maintaining backward compatibility with the default constructor.

RussellSpitzer commented:
Title of this pr looks incorrect? This is a new feature and not a fix. The PR also would require some tests to prove the new parameter is working as expected.

jangalasriramd7 commented:
I have changed the PR title.

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java:
```diff
@@ -47,12 +47,19 @@ public abstract class SizeBasedDataRewriter extends SizeBasedFileRewriter<FileSc
+
+  // Default constructor
+  public SizeBasedDataRewriter(Table table) {
+    this(table, 0.3); // Providing default deleteRatioThreshold
```
Rather than having a magic number here we should keep DELETE_RATIO_THRESHOLD and rename it DEFAULT_DELETE_THRESHOLD and use that here

RussellSpitzer commented on core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java:
```diff
@@ -47,12 +47,19 @@ public abstract class SizeBasedDataRewriter extends SizeBasedFileRewriter<FileSc
   private int deleteFileThreshold;
 
-  protected SizeBasedDataRewriter(Table table) {
+  protected SizeBasedDataRewriter(Table table, double deleteRatioThreshold) {
```
Shouldn't need this in the constructor

jangalasriramd7 commented:
@RussellSpitzer Addressed your comments:
- Replaced the magic number with DEFAULT_DELETE_THRESHOLD
- Moved deleteRatioThreshold to operate as an option
Let me know if further changes are needed!

nastra commented on core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java:
```diff
@@ -47,12 +47,13 @@ public abstract class SizeBasedDataRewriter extends SizeBasedFileRewriter<FileSc
-  protected SizeBasedDataRewriter(Table table) {
-    super(table);
+  // Default constructor
+  public SizeBasedDataRewriter(Table table) {
```
I don't think these changes here are necessary

nastra commented on core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java:
```diff
@@ -67,6 +68,8 @@ public Set<String> validOptions() {
   public void init(Map<String, String> options) {
     super.init(options);
     this.deleteFileThreshold = deleteFileThreshold(options);
+    this.deleteRatioThreshold = PropertyUtil.propertyAsDouble(
```
we also need to validate that `0.0 > threshold <= 1.0`. That would also require some additional unit tests that check for valid/invalid parameters

nastra replied:
also `this.deleteRatioThreshold` needs to be used in `tooHighDeleteRatio()`

nastra commented on core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java:
```diff
@@ -67,6 +68,8 @@ public Set<String> validOptions() {
     super.init(options);
     this.deleteFileThreshold = deleteFileThreshold(options);
+    this.deleteRatioThreshold = PropertyUtil.propertyAsDouble(
+            options, "delete-ratio-threshold", DEFAULT_DELETE_THRESHOLD);
```
`delete-ratio-threshold` should be a static final (similar to `DELETE_FILE_THRESHOLD`) and have a proper javadoc documentation so that it's clear what this parameter is used for

nastra commented on core/src/test/java/org/apache/iceberg/actions/TestSizeBasedRewriter.java:
```diff
@@ -76,6 +76,27 @@ public void testSplitSizeLowerBound() {
+    SizeBasedDataFileRewriterImpl rewriter = new SizeBasedDataFileRewriterImpl(table);
+
+    Map<String, String> options = ImmutableMap.of(
+            SizeBasedDataRewriter.DELETE_FILE_THRESHOLD, "5"
```
`DELETE_FILE_THRESHOLD` is different from the delete ratio and is for something else

jangalasriramd7 commented:
Could you please help me out in resolving the error.

This build must be run with JDK 11 or 17 or 21 but was executed with JDK 23

![image](https://github.com/user-attachments/assets/b27f9234-5e22-4f62-9a3e-d8781b4328c0)
![image](https://github.com/user-attachments/assets/1eb93617-d03d-41ca-8ee5-ede755a21219)

RussellSpitzer commented:
> Could you please help me out in resolving the error.
> 
> This build must be run with JDK 11 or 17 or 21 but was executed with JDK 23
> 

It looks like you may have an issue with your gradle plugin, what have you got setup there?
